[1단계 - 출석 구현] 밍트(김명지) 미션 제출합니다.

Starlight258 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - [작성한 프롤로그 글](https://prolog.techcourse.co.kr/studylogs/3965)


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
설계를 잘 수행했지만, 후에 시간이 부족해서 리팩토링할 부분이 있기 때문입니다.

  
## 어떤 부분에 집중하여 리뷰해야 할까요?
책임 분리가 잘 되었는지, 설계 위주로 봐주시면 감사드리겠습니다.

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->
### 10줄 제한 요구사항
```java
    private void process(final Crews crews, final Command command) {
        if (command.equals(Command.CHECK_ATTENDANCE)) {
            checkAttendance(crews);
        }
        if (command.equals(Command.MODIFY_ATTENDANCE)) {
            modifyAttendance(crews);
        }
        if (command.equals(Command.CHECK_ATTENDANCE_BY_CREW)) {
            checkAttendanceHistoryByCrew(crews);
        }
        if (command.equals(Command.CHECK_DISMISSAL_CREW)) {
            checkDismissalCrews(crews);
        }
    }
```
- 10줄 제한 요구사항을 지키기 위해 충분히 명확함에도 불구하고 메서드를 분리하는 경우가 있었습니다. 페어와 함께 이 메서드를 분리하면 좋을지, 10줄 제한 요구사항을 지키지 않을지 토론했습니다.

우선 요구사항이기 때문에 지켰지만, 10줄 제한 요구사항이 어떠한 뜻을 가지고 있는지 리뷰어님의 생각이 궁금합니다!

리뷰 감사합니다! 🙇‍♀️

unh6unh6 committed:
docs(README.md): 요구사항 기능 명세 작성

unh6unh6 committed:
feat(Crew): 출석 확인 기능 구현

unh6unh6 committed:
feat(input): 닉네임, 등교 시간 입력 기능 구현

unh6unh6 committed:
feat(ResultView): 출석 기록 출력 기능 구현

unh6unh6 committed:
feat(AttendanceType): 출석 종류 계산 기능 구현

unh6unh6 committed:
fix(AttendanceType): 출석 시간 이전의 출석도 지각 또는 결석으로 계산되는 문제 수정

unh6unh6 committed:
feat(Crews): 크루 조회 기능 구현

unh6unh6 committed:
feat(Campus): 등교일이 아닌 경우 예외 처리

unh6unh6 committed:
feat(Campus): 등교일이나 등교 시간이 아닌 경우 예외 처리

unh6unh6 committed:
feat(util): 시간 형식 검증 기능 구현

unh6unh6 committed:
refactor(TimeFormatter): 정적 메서드로 변경

unh6unh6 committed:
fix(InputView): 잘못된 메서드 순서 변경

unh6unh6 committed:
refactor(TimeFormatter): LocalDateTime에서 LocalDate로 매개변수 수정

- 필요한 데이터만 받기 위해서이다

unh6unh6 committed:
feat(InputView): 출석 수정 기능 입력 및 검증 구현

unh6unh6 committed:
docs(README): 구현 기능 체크

unh6unh6 committed:
feat(Crew): 출석 수정 기능 구현

unh6unh6 committed:
feat(StringParser): 유효하지 않은 12월 일자 검증 구현

unh6unh6 committed:
feat(Crew): 과거 날짜 수정 시 예외 기능 구현

unh6unh6 committed:
feat(ResultView): 출석 수정 결과 출력 기능 구현

unh6unh6 committed:
feat(Crew): 크루별 출석 기록 확인 기능 구현

unh6unh6 committed:
refactor(Crew): 불필요한 nickname 필드 제거

unh6unh6 committed:
refactor(Crew): 생성자로 출석 기록 필드를 받도록 수정

unh6unh6 committed:
feat(AttendanceType): 출석, 지각, 결석 횟수 계산 기능 구현

unh6unh6 committed:
feat(SubjectType): 제적 대상자 결정 기능 구현

unh6unh6 committed:
feat(ResultView): 크루별 출석 기록 출력 기능 구현

unh6unh6 committed:
test(Crews): 크루별 출석 기록 기능 테스트

unh6unh6 committed:
feat(Crews): 제적 위험자 정렬 기능 구현

unh6unh6 committed:
feat(Initializer, FileReader): 파일에서 크루와 출석 정보를 불러오는 기능 구현

unh6unh6 committed:
feat(AttendanceController): 입력에 따른 분기 처리 구현

unh6unh6 committed:
fix(ResultView): 잘못된 시간 출력 형식 수정

unh6unh6 committed:
fix(Crews): 제적 대상자가 아닐 경우 출력하지 않도록 수정

unh6unh6 committed:
fix(AttendanceController): 등교일이 아닐 경우 입력받지 않고 바로 예외 발생하도록 수정

unh6unh6 committed:
refactor(InputView): 개행 추가

unh6unh6 committed:
refactor(All): 메서드 10줄 지키기 위해 리팩터링

unh6unh6 committed:
docs(README): 완료한 기능 체크

EastHShin commented on src/main/java/model/AttendanceType.java:
```diff
@@ -0,0 +1,58 @@
+
+public enum AttendanceType {
+
+    출석(Duration.ofMinutes(0)),
```
![image](https://github.com/user-attachments/assets/67c5a423-3866-4b8b-adfb-b8e1cf48f7ac)


오... enum 을 한글로 정의하셨군요!
어떤 장점이 있으셨나요?

Starlight258 replied:
상수명을 한글로 정의하면 가독성도 좋아지고 의도를 빠르게 파악할 수 있는 장점이 있습니다.
그런데 `intellij`에서 `Non-ASCII `경고가 떠서 알아봤더니 호환성 문제가 발생할 수 있다고 해요.

혹시 상수 또는 변수명을 한글로 작성하는 것에 대해 어떻게 생각하시나요? 외국인 개발자와의 협업이나 호환성 문제를 고려하여 영어로 통일하는 것이 좋을까요 ?

EastHShin replied:
일단 기본적으로 Non_ASCII 이기 때문에 호환성 문제, 직렬화 호환 문제 등 여러 문제가 발생할 수 있기 때문에 사용하지 않는 것이 좋다고 생각합니다.
말씀주신 가독성 조차 더 좋은진 잘 모르겠어요. 영어로만 코드를 읽다가 갑작스레 주석이 아닌 변수명에서 한글이 나오면 조금 익숙치 않을 것 같아서요ㅎㅎ

Starlight258 replied:
그렇군요, 변수명을 영어로 쓰다가 enum 부분만 한글인 것이 일관성이 없다고도 느껴지네요!
반영하겠습니다! 

57c2cc13f1970c3d9f4cabbae5f4e4426ab4a68e

EastHShin commented on src/test/java/util/TimeFormatterTest.java:
```diff
@@ -0,0 +1,43 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class TimeFormatterTest {
```
실패하는 테스트가 있네요~ PR 혹은 code push 전 테스트가 정상 동작 하는지 항상 확인하는 습관을 가지면 좋습니다!

Starlight258 replied:
<img width="448" alt="image" src="https://github.com/user-attachments/assets/31dd8f97-1f47-4b1a-961a-4c8accb0199e" />

헉 그런가요, 다른 환경에서 확인해봐도 테스트가 모두 통과되었는데.. 혹시 에러 메세지 공유해주실 수 있나요?

EastHShin replied:
![image](https://github.com/user-attachments/assets/72970d8f-d4d9-4bbb-b912-25d295c280f0)

저는 이렇게 나오네요! 관련해서 디버깅을 해보니, DatetimeFormatter 에서 Locale 설정을 해주지 않으면 시스템 언어를 따라가서 나타나는 현상이네요ㅎㅎ 제 노트북 시스템 언어가 영어라 위와 같은 현상이 발생했어요
한글을 출력하는 요구사항에서는 위와 같은 문제가 발생할 수 있으니 참고만 해두시면 좋을 것 같네요!

Starlight258 replied:
아 그렇군요! 언어 설정에 따라 달라지는 것을 간과했네요! 
환경에 따라 테스트가 깨질 수 있다는 점 기억하겠습니다! 감사합니다!

24c15c80623e6e39b101318f48a5fa1ea9d5b558

EastHShin commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,120 @@
+
+    private void process(final Crews crews, final Command command) {
+        processAttendance(crews, command);
+        if (command.equals(Command.CHECK_ATTENDANCE_BY_CREW)) {
+            checkAttendanceHistoryByCrew(crews);
+        }
+        if (command.equals(Command.CHECK_DISMISSAL_CREW)) {
+            checkDismissalCrews(crews);
+        }
```
입력에 대한 조건문을 만족해도 다음 조건문으로 또 진입하네요~

Starlight258 replied:
`early return`을 사용하는 것이 좋겠네요! 감사합니다!

fae0fd63d2bc68be1f5d1d2c52947d7e7ba64e83

EastHShin commented on src/main/java/model/Command.java:
```diff
@@ -0,0 +1,25 @@
+
+import java.util.Arrays;
+
+public enum Command {
```
입력값에 따른 진행 상황을 저의한 enum 클래스인데, model 패키지에 있는게 적절한 지 모르겠어요! view 에 있어도 되지 않을까요?

Starlight258 replied:
확실히 `Command`는 모든 사용자에게 똑같이 보여주는 부분이라, `model`보다는 `view`에 더 알맞네요!
단순히 객체니까 `model` 패키지에 두었는데, 이번 기회에 `mvc` 패턴에 대해 다시 공부하게 되었어요! 감사합니다. 🎶

c1adc66fb31682c57814174795ed1849cbd2035f

EastHShin commented on src/main/java/model/Crew.java:
```diff
@@ -0,0 +1,55 @@
+
+    public List<LocalDateTime> getAttendanceHistory(final LocalDate todayDate) {
+        int today = todayDate.getDayOfMonth();
+        return IntStream.rangeClosed(1, 31)
+                .filter(attendance::containsKey)
+                .filter(day -> day < today)
+                .mapToObj(attendance::get)
+                .toList();
```
1~31 을 모두 반복할 필요가 있을까요? 어떻게하면 더 효율적으로 개선해볼 수 있을까요?

Starlight258 replied:
어제 날짜까지만 반복하면 되므로 아래와 같이 수정하면 더 명확해질 것 같아요.

```java
public List<LocalDateTime> getAttendanceHistory(final LocalDate todayDate) {
    int today = todayDate.getDayOfMonth();
    return IntStream.range(1, today)
            .filter(attendance::containsKey)
            .mapToObj(attendance::get)
            .toList();
}
```

86e68e18694c0d1294e086d4a8962de4e3569753

EastHShin replied:
```
int today = todayDate.getDayOfMonth();
return attendance.entrySet().stream()
                .filter(it -> it.getKey() < today)
                .map(Entry::getValue)
                .collect(Collectors.toList());
```
위와 같이 작성하는게 더 효율적일 것 같아요!
1부터 today 까지 반복하기 보단, map 에 존재하는 key들만 순회하면 되니까요

Starlight258 replied:
그렇네요! 이미 출석 가능한 날짜를 순서에 맞게 저장해두었으니, 순서가 결정된 상태네요! 반영하겠습니다 !

09d6137ab8ee49c1356d5f1ee7c436d41e801457

EastHShin commented on src/main/java/model/Crews.java:
```diff
@@ -0,0 +1,47 @@
+        return crews.get(nickname);
+    }
+
+    public List<DismissalCrewDto> findDismissalCrewDtos(final LocalDate todayDate) {
```
도메인에서 DTO를 알 필요가 있을까요? 밍트가 생각하는 DTO의 역할은 무엇인가요?

Starlight258 replied:
`dto`는 **계층간 데이터 교환**을 위해 사용하는 객체입니다. `controller`는 `view`와 `model의 데이터`를 주고받을 때 `dto`를 사용합니다. `dto`는 요구사항에 의해 자주 바뀔 수 있기 때문에, 여러 곳에서 참조되는 `entity`가 `dto`를 의존하면(알면) 안됩니다.
즉, 도메인은 비즈니스 로직에만 집중해야하며, 표현 계층의 관심사가 도메인에 침투하면 변경에 취약할 수 있습니다.

해당 사항 반영하겠습니다!

6d99aa1ce00ce1e23f4f719101d1bb90f4311b76

Starlight258 replied:
리팩토링하면서 궁금한 점이 있어 남깁니다!

## 질문 1. 연산 횟수 문제
### 4번 기능 : 제적 위험자 확인 로직
- 제적 위험자 확인하기
1. 크루 집합(`CrewHistories`)에서 크루별 출석 기록(`CrewHistory`)을 조회한다.
2. `CrewHistory`에서 출석 타입(`AttendanceType`) 별 횟수를 계산한다.
3. 출석 타입별 횟수를 통해 `SubjectType`에서 제적 대상자인지 확인한다.

- 제적 위험자일 경우 횟수 출력하기
1. `CrewHistory`에서 출석 타입별 횟수를 계산하여 반환한다.

위 기능을 실행하면 **CrewHistory에서 출석 타입별 횟수를 계산하는 로직**이 2번 실행되는데, 개선할 수 있는 방법 있을까요?

처음에는 출석 타입별 횟수(출석/지각/결석)를 `CrewHistory` 객체의 필드로 두려고 했습니다. 그런데 출석 수정 기능으로 인해 출석이 달라질때마다 매번 재계산해야하고, 이미 존재하는 필드인 `출석 기록`에 대한 부차적 데이터라는 생각이 들어 두지 않았어요.

객체지향으로 역할을 분배하니 매번 연산을 수행하게 되는 상황이 이전에도 발생한 적이 있는데, 제 설계 역량이 아직 부족한 것인지 객체지향이 가진 한계인지 궁금해요!

## 질문 2. 출력시 정렬에 대한 책임
```
제적 위험자 조회 결과
- 빙티: 결석 3회, 지각 2회 (면담)
- 이든: 결석 2회, 지각 4회 (면담)
- 쿠키: 결석 2회, 지각 2회 (경고)
- 빙봉: 결석 1회, 지각 5회 (경고)
```
제적 위험자는 주어진 기준에 따라 정렬해야하는데요, 지금은 DTO(`DismissalCrewDto`)에서 수행하고 있습니다.
그런데 출력이니 `view`에서 도메인의 데이터를 받아 정렬해야하나 고민이 들었어요. 이스트의 생각이 궁금해요!

EastHShin replied:
>질문 1. 연산 횟수 문제
 
밍트가 작성해주신 로직을 보면, 제적 대상자를 찾아오는 로직에서 List<CrewHistory> 를 반환하는데요. 이 CrewHistory에는 출석 시간이 들어 있습니다. 제적 대상자를 찾는데, 출석시간을 불러온 뒤에 이를 DTO 에서 또 계산해주어 제적 대상자를 확정 짓는 로직으로 반복중이에요. 제적 대상자를 찾는 로직에서 출석 시간을 반환할 필요가 있을까요?
출석기록을 통해 제적 위험자만 받아오면 되지 않을까요? 그럼 밍트가 고민한 부분도 해결이 될 것이라고 생각해요.

>질문 2. 출력시 정렬에 대한 책임

네, DTO는 데이터를 전달하는 객체일 뿐이기 때문에 정렬의 책임을 갖기보단, View에서 이를 수행하는 것이 맞다고 생각합니다.
정렬은 출력 요구사항에 해당할 뿐이지 비즈니스 로직과는 연관이 없다고 생각해요.

Starlight258 replied:
> 1. 질문 1. 연산 횟수 문제

그렇네요! 생각해보니 이미 `출석 기록`으로부터 `횟수`를 세었는데, `횟수` 값을 활용하지 않고 다시 `출석 기록`을 넘기네요!
`dto`를 사용하지 않고 `닉네임과 출석 타입별 횟수인 map`을 함께 넘기는 방법이 떠오르지 않아서 그랬던 것 같아요.
그래서 `dto`에서 로직을 중복으로 수행하게 되었군요..

`출석 타입별 횟수`인 `map`은 `AttendanceCounter`라는 이름으로 객체를 만들고, 제적 대상자를 찾는 로직에서 `Map<String, AttendanceCounter> `(닉네임, 출석타입별 횟수)를 반환하도록 수정했습니다!

`stream`으로 한번에 처리하려고 하다보니 그러한 문제가 발생한 것 같아요, 앞으로는 도메인에서 최대한 로직을 수행하고, `dto`에는 로직을 최대한 덜 사용하겠습니다!

0ebccfe1a5f92b865ba759104dde135c81872b6c

Starlight258 replied:
> 질문 2. 출력시 정렬에 대한 책임

그렇군요! `record`는 단순 데이터 홀더이니 출력 로직을 수행하기에는 적합하지 않군요! `view`에서 수행하도록 수정하겠습니다!

cc2fd5303e24c24cb7d1c8a17098f7fe0db9de4f

EastHShin commented on src/main/java/model/Initializer.java:
```diff
@@ -0,0 +1,69 @@
+
+    public Crews initialize() {
+        Map<String, Crew> crewsMap = new HashMap<>();
+        List<String> lines = FileReader.readFile();
```
현재는 csv 파일을 통해 출석정보를 초기에 적재하는 요구사항이지만, 다른 방법으로 읽어오도록 요구사항이 변경되면 어떻게 될까요?
외부 저장소에서 읽어 오도록 바뀐다면 initializer 로직 자체가 영향이 받을 것 같아요. 그러한 외부 입력 요구사항에 영향을 받지 않으려면 어떻게 개선해 볼 수 있을까요?

Starlight258 replied:
파일을 읽어오는 부분을 `Initializer` 바깥에 두고, `Initializer`는 읽어온 데이터를 인자로 받아 처리하면 좋을 것 같습니다 !

8e66e09b32677015d048fbf22d97562ce4728fa9

Starlight258 replied:
리팩토링하다 궁금한 점이 있어 남깁니다!

### `Initializer`에 대한 고민
모든 크루에 대해 12월에 대한 출석 기록을 초기화해주어야하는데요, 제 코드는 아래와 같이 동작합니다.

1. `Initializer`에서 12월 첫 운영 날짜부터 어제 일자까지 결석으로 정의한 `Map<Integer, LocalDateTime>`를 생성합니다.
2. `CrewHistory` 생성시 해당 출석 기록(`Map<Integer, LocalDateTime>`)을 저장합니다.
3. 파일에서 출석 기록을 읽어와 크루별로 출석을 수정합니다.

여기서 초기화하는 과정을 `CrewHistories`에서 수행할지, 지금처럼 `Initializer`에 할지 페어와 토론을 했었어요.

`CrewHistories`에서 초기화를 하면, 출석 기록에 대한 로직이 모이게 되어 응집도가 높아진다는 생각이 들었습니다.
그런데 페어는 모든 크루에 대해 같은 초기화 동작을 하니 `Initializer`에서 수행하는 것이 나을 것 같다는 이야기를 했었어요.

결론적으로 `CrewHistories`가 과도한 책임을 가지게 될까봐 `Initializer`에서 초기화를 수행하는 것으로 결론을 짓게 되었는데요, 여전히 고민이 됩니다. 응집도 측면에서 `CrewHistories`에 둘지 고민이 되는데, 어떻게 생각하시나요?


### 프로그램 운영시 문제 
작성한 프로그램은 프로그램을 실행할 때마다 어제 날짜까지의 기록을 `결석`으로 초기화하고, 파일에서 출석 기록을 `업데이트`하는 형식인데요, 만약 이 프로그램이 꺼졌다 켜지게 되면 출석 확인, 출석 수정 기록이 사라지니 파일 쓰기 기능이 필요할 것 같아요.

만약 프로그램을 계속 켜둔다면, 출석 초기화가 제대로 안되어 버그 상황으로 이어지게 되는데요, 이스트가 생각하시기에 출석 시스템은 껐다 켰다 하는 시스템인지, 웹서버처럼 계속 운영되는 시스템이라고 생각하시나요? 그에 따라 설계가 수정될 것 같아서요, 저는 전자같긴 합니다!

EastHShin replied:
>Initializer에 대한 고민

저였다면, Initializer(CrewInitializer) 를 통해 출석기록을 생성해주었을 것 같아요.
CrewHistories 는 기능과 관련된 비즈니스 로직 (제적 학생 필터링 등)의 책임을 가지고 있고, 외부 파일에서 읽어 출석기록을 가져오는 것에 대한 책임 자체는 CrewIntializer 가 더 적합하지 않을까 생각이 드네요. CrewHistories 라는 도메인 객체가 초기 데이터 생성에 관한 것을 과연 알고 있어야 할까? 라고 생각해요. CrewHistories 의 핵심적인 책임은 CrewHistory를 이용해 제적 학생을 찾는 행위인데, 이 행위 외에 초기 데이터 생성은 불필요한 결합인 것 같아요.

>프로그램 운영시 문제

기본적으로 상용 소프트웨어는 계속 운영되는 시스템으로 상정하고 설계를 해야 한다 생각해요. 출석 시스템을 매번 껐다 켰다 할 수는 없으니, 출석 시간만을 가지고 기능 사용시마다 결석, 지각등의 계산을 하도록 할 것 같네요. 
그런데, 밍트가 구현하신 로직이 제가 말한 방식 아닌가요? 현재 제가 보고있는 밍트의 로직으로는 크루별 출석 시간 목록을 가지고 있고, 이를 기반으로 런타임에 출석 여부를 계산하는 것으로 보여서요. 파일 쓰기 기능을 도입하더라도 말씀해주신 출석 초기화가 안되는 문제는 없지 않나요?
제가 질문을 잘 이해하지 못한 것일수 있으니, 추가적인 설명이나, 참고가 될만한 코드에 코멘트 남겨주시면 답변드릴게요!

Starlight258 replied:
> Initializer에 대한 고민

그렇군요! 초기 데이터 생성 책임이 객체의 핵심 책임은 아니니 따로 분리해도 좋군요! 감사합니다!

31ae02eddb812bb4b99a4cbcb35c0848a79c0ab5

> 프로그램 운영시 문제

저는 껐다 키는 시스템이라고 생각하고 구현을 했는데, 만약 계속 운영되는 시스템이라면 아래와 같은 문제가 발생해요.

만약 월요일에 프로그램을 켜둔 상태로, 제가 월, 화, 수 출석하지 않는다면 결석 수가 3 증가되어야하는데 월요일에 `Initializer`에서 **어제 날짜(지난주 금요일)까지만 결석으로 초기화**해주었기 때문에 출석 확인을 하지 않으면 결석 수가 증가되지 않은 문제가 있어요.

따라서 켰다 끄지 않고 계속 운영되는 시스템이라면, 초기화 시에 **미리 한달 전체의 캠퍼스 운영 날짜를 결석으로 초기화**하거나 출석 타입 별 횟수를 계산할 때 **한 달을 일자별로 순회**하여야 할 것 같아요.

설계를 수정하는 것이 좋을까요?

EastHShin commented on src/main/java/model/Initializer.java:
```diff
@@ -0,0 +1,69 @@
+
+    private Map<Integer, LocalDateTime> createInitialAttendance() {
+        Map<Integer, LocalDateTime> initialAttendance = new HashMap<>();
+        LocalDate now = AttendanceController.getTodayDate();
```
모델 패키지에 두셨는데, 컨트롤러를 의존하고 있네요! 이 경우에는 어떤 단점이 있을까요?

Starlight258 replied:
`MVC` 패턴에서 `모델`은 `컨트롤러`에 의존하면 안됩니다. `컨트롤러`(표현 계층)이 변경되면 `모델`(도메인 계층) 또한 변경되기 때문입니다. 모델은 핵심 비즈니스 로직을 담고 있고, 여러 곳에서 참조되고 재사용되어야하므로 외부 변경으로부터 보호해야합니다.

1c43281575d8b366730e7e6140039a2e95c9a022

EastHShin commented on src/main/java/model/SubjectType.java:
```diff
@@ -0,0 +1,42 @@
+        return absentCount + lateCount / CONVERTED_ABSENT_UNIT;
+    }
+
+    public static int compare(final SubjectType type1, final SubjectType type2) {
```
현재의 type1, type2 변수명으로는 무엇을 어떻게 비교하는지, 기준이 무엇인지 한눈에 알기가 어려워 보여요. 적절한 변수명으로 수정해봅시다!

Starlight258 replied:
`baseType`, `comparedType`으로 수정하고, `Comparator`를 정의하여 가독성을 높이도록 수정했습니다!

```java
private static final Comparator<SubjectType> COMPARATOR = (baseType, comparedType) ->
            Integer.compare(comparedType.threshold, baseType.threshold);
```

e89b17036803c750022807d083846c089188e28b

EastHShin commented on src/main/java/model/SubjectType.java:
```diff
@@ -0,0 +1,42 @@
+        return Arrays.stream(SubjectType.values())
+                .sorted((subjectType1, subjectType2) -> Integer.compare(subjectType2.threshold, subjectType1.threshold))
+                .filter(subjectType -> totalAbsentCount >= subjectType.threshold)
+                .findFirst()
```
현재의 코드에서는 관련이 없지만, 병렬 스트림을 사용한다면 findFirst() 와 findAny() 에는 어떤 차이가 있을까요?

Starlight258 replied:
병렬 스트림에서 `findFirst()`는 `Stream`의 가장 앞 요소를 반환하지만, `findAny()`는 가장 먼저 찾은 요소를 반환합니다.
즉, `findAny()`는 `Stream`의 뒤쪽에 있는 원소가 리턴될 수 있습니다.

만약 스트림 요소를 정렬(`sorted()`)하지 않았다면, `enum` 순서에 의존하기 때문에 병렬 처리시에 문제가 있을 수도 있겠네요! 병렬 처리에 대해 잘 몰랐는데 한번 공부해보겠습니다! 감사해요!

EastHShin commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,120 @@
+        }
+    }
+
+    private void processAttendance(Crews crews, Command command) {
```
```private void process(final Crews crews, final Command command) {
        if (command.equals(Command.CHECK_ATTENDANCE)) {
            checkAttendance(crews);
        }
        if (command.equals(Command.MODIFY_ATTENDANCE)) {
            modifyAttendance(crews);
        }
        if (command.equals(Command.CHECK_ATTENDANCE_BY_CREW)) {
            checkAttendanceHistoryByCrew(crews);
        }
        if (command.equals(Command.CHECK_DISMISSAL_CREW)) {
            checkDismissalCrews(crews);
        }
    }
```
> 10줄 제한 요구사항을 지키기 위해 충분히 명확함에도 불구하고 메서드를 분리하는 경우가 있었습니다. 페어와 함께 이 메서드를 분리하면 좋을지, 10줄 제한 요구사항을 지키지 않을지 토론했습니다.
우선 요구사항이기 때문에 지켰지만, 10줄 제한 요구사항이 어떠한 뜻을 가지고 있는지 리뷰어님의 생각이 궁금합니다!

지금은 입력을 5개만 받아 처리하고 있지만, 받을 수 있는 입력이 늘어날 경우엔 어떨까요?
process 메서드가 지금은 몇 줄 안되지만 늘어난 입력 커맨드에 따라 메서드가 커지지 않을까요?
메서드가 크다는건 그 메서드가 하는 일이 많은건 아닌지 의심해보아야 합니다.
10줄 제한사항이라는 것을 둔 이유는 단순히 가독성을 위한 것이 아니라 객체지향적인 관점에서 한 행위가 너무 크거나 여러가지의 역할을 담당하고 있는건 아닌지 10줄 제한을 통해 의심을 하는 것이라고 생각해요.

그런 의미에서 이 메서드는 어떻게 확장성 있게 리팩터링 할 수 있을까요?! 감이 안잡히면 추가적으로 코멘트나 DM으로 질문 주셔도 됩니다!

Starlight258 replied:
받을 수 있는 입력이 늘어나는 경우에 확실히 메서드가 길어지는 문제가 있네요. 
고민해보다가 함수형 인터페이스 중 `Consumer`를 사용하여 리팩토링을 해보았어요.

```java
public class AttendanceController {

    private final Map<Command, Consumer<CrewHistories>> commands = Map.of(
            Command.CHECK_ATTENDANCE, this::checkAttendance,
            Command.MODIFY_ATTENDANCE, this::modifyAttendance,
            Command.CHECK_ATTENDANCE_BY_CREW, this::checkAttendanceHistoryByCrew,
            Command.CHECK_DISMISSAL_CREW, this::checkDismissalCrews
    );
    
       private void process(final CrewHistories crewHistories, final Command command) {
        Consumer<CrewHistories> consumer = commands.get(command);
        consumer.accept(crewHistories);
    }
```
훨씬 나아졌는데, `Controller`에 `commands`가 존재하는 것이 맞는지, 실행 흐름의 책임을 두어야할지 고민이 됩니다. 확장성 있게 따로 객체로 두거나 `enum`으로 실행 흐름 로직을 옮기고 싶은데, `controller`의 메서드를 옮겨야해서 (ex) `checkAttendance`) 고민이 되어요. 이스트의 생각이 궁금해요!

0c040ca522e519671dd06d0715067021febae102

EastHShin replied:
현재는 컨트롤러에 각 기능에 대한 모든 실행 흐름을 두고 있는데, 각 기능마다 클래스를 분리해 실행 흐름을 넘겨주는 방법도 괜찮아 보여요!
관련해서 [Command Pattern](https://www.geeksforgeeks.org/command-pattern/) 을 학습해보시면 좋을 것 같아요.

Starlight258 replied:
`Command` 패턴을 사용하여 리팩토링해보았어요~
```java
public class FindByCrewCommand implements Command {

    private final InputView inputView;
    private final TodayClock todayClock;
    private final ResultView resultView;

    public FindByCrewCommand(final InputView inputView, final TodayClock todayClock, final ResultView resultView) {
        this.inputView = inputView;
        this.todayClock = todayClock;
        this.resultView = resultView;
    }

    @Override
    public void execute(final CrewHistories crewHistories) {
        String nickname = inputView.readNickname();
        CrewHistory crewHistory = crewHistories.findCrewByNickname(nickname);

        LocalDate todayDate = todayClock.getTodayDate();

        List<LocalDateTime> attendanceHistory = crewHistory.getAttendanceHistory(todayDate);
        AttendanceCounter attendanceCounter = crewHistory.countAttendanceType(todayDate);

        resultView.printAttendanceHistoryResultByCrew(nickname, attendanceHistory, attendanceCounter);
    }
}
```

그런데 기능마다 입력 매개변수와 반환 타입이 달라서 `view`를 의존하게 되었어요~
따라서 `command` 패키지를 `domain`에서 `controller`로 두었습니다.

b8d3acfa1f9bf0dacda09668254ca2312e643f24

EastHShin commented on src/main/java/model/Crew.java:
```diff
@@ -0,0 +1,55 @@
+
+public class Crew {
+
+    private final Map<Integer, LocalDateTime> attendance;
```
현재는 12월에 한정된 요구사항이지만, 정식적으로 이 시스템이 사용된다면 어떻게 될까요?
12월에 한정된 날짜를 key로 출석 시간을 받고 있는데, 다른 월까지 확장된다면 Crew에 있는 모든 로직이 영향을 받을 것 같아요.
물론 현재의 요구사항에서는 알맞은 구현을 해주셨지만, 설계에 관한 리뷰를 요청하셔서 참고 사항 정도로만 피드백 드려요!

Starlight258 replied:
그렇겠네요! 사실 처음에는 `List<LocalDateTime>`으로 출석 기록을 가지고 있다가, 수정시에 출석 기록을 `일자`로 접근하여서 고민끝에 `Map`을 사용하게 되었어요. 만약 12월 외에 다른 달도 지원하도록 한다면 `Map<LocalDate, LocalDateTime>`을 사용하는 것이 좋겠네요!

우선 `YAGNI` 원칙에 따라 위와 같이 구현했는데, 저도 확장성 측면에서 많이 고민한 부분이라서요. 이스트라면 어떤 자료구조를 사용하실건지 생각이 궁금해요!

EastHShin replied:
YAGNI 원칙을 언급해주셔서 잠깐 얘기하지면... 이런 고민은 평소에 저도 참 많이 하는데요. 과연 어느정도 까지 확장성을 생각하고 개발해야 할 지, 어느 선이 적절할지 고민이 많이 되는 것 같아요.
기능적으로 확장될 수 있는 여지는 항상 있고, 그렇기 때문에 확장성 있게 설계하려고 하고, 그 점이 효과적으로 작동한 적이 많았던 것 같아요. 그리고 이를 판단하는 기준에는 
- 경험적으로 이 기능이 확장될 여지가 많은가?
- 확장성있게 설계하는데 훨씬 많은 리소스가 드는가?
- 확장성있게 설계하지 않았는데, 추후 예상했던 대로 기능이 확장된다면 이를 수정하는데 너무 많은 노력이 들지는 않는가?
정도 인 것 같아요.

출석 시스템의 경우에는 만약 미션이 아닌 실제로 시스템을 시범적으로 12월에 운영한다고 했을 때, 정식 시스템이 될 경우 높은 확률로 모든 월에 대해 기능이 확장될테고, 확장성있게 설계를 하는 것이 12월만 고려해 개발하는것에 비해 많은 리소스가 더 들지 않기 때문에 모든 월에 대해 대응할 것 같네요.

이야기가 많이 샌 것 같은데... 결론으로 돌아와서 모든 월, 그 뿐만 아니라 다른 해에 대한 대응 또한 하기 위해서는 LocalDate가 적절할 것 같네요! 혹은, LocalDateTime 리스트를 갖고 있고, 이를 날짜별 groupby 해 사용하는 것도 괜찮을 것 같아요

Starlight258 replied:
피가 되고 살이 되는 조언 감사해요 😁
도메인에 대한 이해도가 중요한 것 같네요!

확실히 출석 시스템은 모든 월에 대한 기능 확장이 필수 불가결하니,,` Map<LocalDate, LocalDateTime>`으로 설계하는 것이 확장성을 고려한 좋은 설계네요!

앞으로도 말씀해주신 부분 고려해서 설계해볼게요! 감사합니다!

5e11e3ce564906773c3d1094812c2a1eb99bb81a

EastHShin commented on src/main/java/model/Crew.java:
```diff
@@ -0,0 +1,55 @@
+
+    public LocalDateTime doModify(final LocalDateTime modifyDateTime, final LocalDate todayDate) {
+        LocalDate modifyDate = LocalDate.from(modifyDateTime);
+        if (modifyDate.isEqual(todayDate) || modifyDate.isAfter(todayDate)) {
```
조건문의 행위를 한 눈에 알아볼 수 있게 메서드 분리를 해볼까요?

Starlight258 replied:
그렇네요! 너무 구체적인 로직이네요! 메서드를 분리해서 추상화 레벨을 맞추도록 할게요! 

79d95158b37f7d7e81b82cbeb77f44e0ec64ee08

EastHShin reviewed (changes requested):
안녕하세요 밍트!

이번 미션 함께하게 된 이스트입니다 😄 
커밋을 보니 TDD 사이클을 경험해보시려고 한 노력이 느껴져요. 미션 구현 하시느라 고생하셨습니다!
코멘트 남겼으니 확인해주시고, 궁금하신 점은 코멘트나 DM 으로 주시면 답변 드리겠습니다 :)

Starlight258 committed:
refactor(AttendanceController): early return 사용

- early return을 사용하여 필요 없는 작업을 수행하는 것을 막는다.

Starlight258 committed:
refactor(Command): Command 객체 view 패키지로 이동

- Command는 사용자 입력에 대해 작동 흐름을 보여주는 view이다.

Starlight258 committed:
refactor(Crew): stream 범위 조정

- filter()를 사용하기보다, 스트림의 요소들을 한달 전체의 일자가 아닌 어제 날짜까지로 범위를 조정한다.

Starlight258 committed:
refactor(Crews): 방어적 복사 수행

Starlight258 committed:
style(All): 포맷팅 및 필요없는 주석 제거

Starlight258 committed:
test(InputValidator): assertThatCode()에서 예외가 발생하지 않는지 확인

Starlight258 committed:
refactor(Crew): 도메인이 dto를 의존하지 않도록 수정

- dto는 도메인에 비해 자주 변경되기 때문에 도메인이 dto에 의존하면 변경에 취약하게 된다.

Starlight258 committed:
refactor(Initializer): 데이터를 읽는 부분을 메서드 인자로 받도록 수정

- 이를 통해 Intializer가 데이터를 읽는 방식(파일, 외부 저장소 등등)에 영향을 받지 않도록 개선한다.

Starlight258 committed:
refactor(Crew): 메서드 추출

Starlight258 committed:
refactor(CrewHistory): 클래스, 메서드명 변경

- 의도가 드러나도록 이름 변경

Starlight258 committed:
feat(TodayClock): TodayClock 인터페이스를 도입하여 확장성을 높임

- 단순히 현재 시간 구현체를 변경함으로써 오늘 날짜를 쉽게 변경할 수 있다.

Starlight258 committed:
feat(SubjectType): Comparator(비교자)를 사용하여 정렬 로직의 가독성을 높임

Starlight258 committed:
refactor(AttendanceController): 메서드 접근제어자 변경

Starlight258 committed:
refactor(Controller): 입력에 대한 Map 생성

- 제어 흐름에 대한 map을 생성하여 요청된 로직만 수행하도록 한다.

Starlight258 committed:
refactor(CrewHistory): 메서드 분리하여 가독성을 높임

Starlight258 commented:
@EastHShin 리뷰 감사드려요! 덕분에 많이 배웠습니다! 🙇‍♀️ 
개선할 부분 있으면 피드백 부탁드려요!

Starlight258 committed:
refactor(CrewHistory): 메서드 분리

- 메서드가 중복으로 호출되는 것을 막는다.

Starlight258 committed:
refactor(Application): 메서드 분리

EastHShin commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,122 @@
+    private void checkDismissalCrews(final CrewHistories crewHistories) {
+        List<CrewHistory> dismissalCrewHistories = crewHistories.findDismissalCrews(getTodayDate());
+        List<DismissalCrewDto> dismissalCrewDtos = DismissalCrewDto.of(getTodayDate(), dismissalCrewHistories);
+        Collections.sort(dismissalCrewDtos);
```
4번 기능을 이용하면 `UnsupportedOperationException` 이 발생하네요!
왜 발생하는 걸까요? 디버깅 해봅시다!

Starlight258 replied:
리팩토링하면서 `stream`에서 `toList()`를 사용한 것이 원인이었어요! 
`collect(Collectors.toList());`로 변경하여 문제 해결하였습니다! 감사합니다.

EastHShin commented on src/main/java/model/SubjectType.java:
```diff
@@ -0,0 +1,51 @@
+
+    public static boolean isApplicable(final Map<AttendanceType, Integer> result) {
+        SubjectType subjectType = from(result);
+        return !subjectType.equals(SubjectType.해당없음);
```
enum은 equals 가 아닌 `==` 로 비교해도 괜찮습니다. 왜 `==` 를 사용해도 되는 걸까요?

Starlight258 replied:
`enum`은 `jvm`에 의해 싱글톤 특성이 보장되므로(오직 하나의 인스턴스만 생성되므로) 참조가 동일하기 때문에, `==`을 사용하는 것이 좋습니다!
감사합니다!

a32d14dc650b85ab1cc9b64e5206795d00e3e4eb

EastHShin commented on src/main/java/model/Initializer.java:
```diff
@@ -0,0 +1,68 @@
+        this.attendance = createInitialAttendance();
+    }
+
+    public CrewHistories initialize(final List<String> inputs) {
+        Map<String, CrewHistory> crewsMap = new HashMap<>();
+        for (String line : inputs) {
+            loadHistory(line, crewsMap);
+        }
```
initalize 할 때는 크루 닉네임과 출석 시간에 대한 정보만 인자로 받아서 초기화하면 어떨까요? 그럼 파일이 아닌 다른 저장소에서 읽거나, 파일 형식이 바뀌어도 Initializer 로직은 바뀌지 않을수 있지 않을까요?

Starlight258 replied:
그렇네요! `List<String>`은 파일 한줄 한줄 읽는 것과 같으니, `nickname`과 `출석 정보`를 `map`으로 전달하도록 할게요!
그렇게 되면 `Initializer`가 데이터 저장 위치에 상관없이 재사용할 수 있겠네요!

9409712b4e03b3b97e27fe73194f9afbb728e6c5

EastHShin commented on src/main/java/model/AttendanceType.java:
```diff
@@ -0,0 +1,44 @@
+
+    private static LocalTime getStartTime(DayOfWeek dayOfWeek) {
+        if (dayOfWeek == DayOfWeek.MONDAY) {
+            return LocalTime.of(13, 0);
+        }
+        return LocalTime.of(10, 0);
```
매직넘버에 의미를 부여해볼까요?

Starlight258 replied:
반영했습니다! 감사합니다.

6a5cc0aa67ee26511b0086d5243ec850af4818bf

338a996f6e764c94176610c47f507541313d74fe

EastHShin commented on src/test/java/model/CrewHistoriesTest.java:
```diff
@@ -0,0 +1,85 @@
+    void findDismissalCrewsTest() {
+        // Given
+        LocalDate todayDate = LocalDate.of(2024, 12, 10);
+        CrewHistory crewHistory1 = new CrewHistory(hotteok, Map.of(
+                2, LocalDateTime.of(LocalDate.of(2024, 12, 2), DEFAULT_TIME),
+                3, LocalDateTime.of(LocalDate.of(2024, 12, 3), DEFAULT_TIME),
+                4, LocalDateTime.of(2024, 12, 4, 9, 30),
+                5, LocalDateTime.of(2024, 12, 5, 9, 30),
+                6, LocalDateTime.of(2024, 12, 6, 9, 30),
+                9, LocalDateTime.of(2024, 12, 9, 9, 30)
+        ));
+        CrewHistory crewHistory2 = new CrewHistory(mint, Map.of(
+                2, LocalDateTime.of(LocalDate.of(2024, 12, 2), DEFAULT_TIME),
+                3, LocalDateTime.of(LocalDate.of(2024, 12, 3), DEFAULT_TIME),
+                4, LocalDateTime.of(LocalDate.of(2024, 12, 4), DEFAULT_TIME),
+                5, LocalDateTime.of(2024, 12, 5, 9, 30),
+                6, LocalDateTime.of(2024, 12, 6, 9, 30),
+                9, LocalDateTime.of(2024, 12, 9, 9, 30)
+        ));
+
+        CrewHistory crewHistory3 = new CrewHistory(wilson, Map.of(
+                2, LocalDateTime.of(LocalDate.of(2024, 12, 2), DEFAULT_TIME),
+                3, LocalDateTime.of(LocalDate.of(2024, 12, 3), DEFAULT_TIME),
+                4, LocalDateTime.of(LocalDate.of(2024, 12, 4), DEFAULT_TIME),
+                5, LocalDateTime.of(LocalDate.of(2024, 12, 5), DEFAULT_TIME),
+                6, LocalDateTime.of(LocalDate.of(2024, 12, 6), DEFAULT_TIME),
+                9, LocalDateTime.of(LocalDate.of(2024, 12, 9), DEFAULT_TIME)
+        ));
```
테스트 픽스처를 활용해서 명확하고, 관리하기 쉽게 작성해 볼 수 있지 않을까요?

Starlight258 replied:
그럼 코드 중복을 줄일 수 있겠네요! 감사해요 :)

e0cb42929b2089a975c6e3a7141336b285b24991

EastHShin reviewed (changes requested):
안녕하세요 밍트~

피드백 반영해주신 부분 확인했습니다!
질문에 대한 답변과 추가적인 코멘트 남겨두었으니 확인해주세요~

Starlight258 committed:
refactor(AttendanceType, SubjectType): enum 상수명 영어로 수정

- 호환성을 고려하여 영어로 수정한다.

Starlight258 committed:
fix(TimeFormatter): 한글로 수정

- 예상 출력을 한글로 출력되도록 고정한다.

Starlight258 committed:
refactor(CrewHistory): stream 효율적으로 수정

- key로 조회하는 것보다, 내부 요소들을 순회하도록 수정하면 더 명확하다.

Starlight258 committed:
fix(DismissalCrewDto): 수정할 수 있는 리스트로 반환하도록 수정

- toList()는 변경할 수 없는 view(unmodifiableList)를 반환하기 때문에, 정렬을 위해서는 사용해선 안된다.

Starlight258 committed:
refactor(AttendanceController): 필요없는 로직 제거

Starlight258 committed:
refactor(AttendanceCounter): 출석 기록별 횟수를 나타내는 AttendanceCounter 도입

- 제적 대상자를 구할 때, 출석 기록별 횟수를 계산하는 로직을 2번 수행하는 문제가 있었다.
- 따라서 출석 기록별 횟수를 객체로 만들어 연산 결과를 반환하도록 수정함으로써 로직이 1번만 수행되도록 수정하였다.

Starlight258 committed:
refactor(ResultView): 정렬 출력 로직 view로 이동

- dto는 단순 데이터 홀더이므로 출력에 사용되는 정렬 로직은 view로 옮긴다.

Starlight258 committed:
refactor(CrewInitializer): Initializer -> CrewInitializer 클래스명 수정

Starlight258 committed:
fix(CrewHistory): 순서를 보장하기 위해 TreeMap 사용

- TreeMap은 key를 오름차순으로 정렬하기 때문에, 날짜별 출석 기록에 적합한 자료구조이다.
- 모든 자료구조에 LinkedHashMap을 사용하는 것보다 안전하게 정렬하여 출력할 수 있다.

Starlight258 committed:
refactor(MenuOption): Command -> MenuOption 클래스명 변경

- command 패턴과 헷갈릴 수 있으므로 클래스명 변경

Starlight258 committed:
refactor(All): domain 패키지 생성

Starlight258 committed:
refactor(AttendanceController): Command 패턴을 사용하여 확장성 제공

Starlight258 committed:
style(Command): 불필요한 개행 제거

Starlight258 committed:
refactor(CrewHistory): 확장성을 위해 map의 key를 LocalDate로 변경

- 출석부 서비스는 한달 이상 지원하도록 확장 가능하므로 key를 LocalDate 타입으로 변경한다.

Starlight258 committed:
refactor(CrewInitializer): 데이터 위치에 상관없이 데이터 초기화하도록 수정

- List<String> inputs은 파일 한 줄 한 줄을 읽으므로, 데이터 저장소에서 가져온다면 사용할 수 없다.
- Map<String, List<LocalDateTime>>을 받도록 하여, 데이터 저장 위치에 상관없이 초기화 로직을 수행할 수 있다.

Starlight258 committed:
refactor(AttendanceType): 매직 넘버 상수로 정의

Starlight258 commented:
@EastHShin 리뷰 감사드립니다! 🙇‍♀️
덕분에 설계 부분 + 구현 부분 모두 많은 인사이트를 얻었습니다!
정리해서 두고두고 기억하겠습니다!

피드백 반영해서 수정했는데, 개선할 부분 있으면 피드백 부탁드려요!

Starlight258 committed:
refactor(All): 패키지 구조 변경

Starlight258 committed:
test(All): TestFixture을 이용하여 코드 중복 제거

Starlight258 committed:
refactor(AttendanceController): 표준 API인 Clock을 사용하도록 변경

- 기존에 직접 만든 TodayClock 인터페이스를 사용하는 것보다는, 표준 API인 Clock을 사용하여 간결하면서 동일한 기능을 제공한다.

Starlight258 committed:
refactor(SubjectType): enum에서 참조값을 바로 비교하도록(==) 수정

- enum은 싱글톤 특성을 가지므로 참조로 바로 비교해도 좋다.

Starlight258 committed:
refactor(Application): 필요없는 코드 제거

Starlight258 committed:
refactor(Campus): 매직 넘버 상수로 분리

Starlight258 committed:
refactor(SubjectType): 메서드 순서 변경

Starlight258 commented on src/main/java/attendance/util/StringParser.java:
```diff
@@ -0,0 +1,46 @@
+    public static LocalDate parseLocalDate(final String input) {
+        int day = parseInt(input);
+        try {
+            return LocalDate.of(2024, 12, day);
```
<img width="638" alt="image" src="https://github.com/user-attachments/assets/a4fab95c-5214-4197-b21b-89216d9b448b" />

~2024년, 12월에 대한 코드가 여기저기 흩어져있는데, 다른 달을 지원하도록 확장성을 고려한다면 enum(상수)로 관리하는 것이 좋을까요?~ 

---

`LocalDate.now(clock)`에서 날짜를 가져와 상수는 `clock` 한 곳에만 존재하도록 수정했습니다! 
05f8814ac8feba444ef35f74605f815f77611fdb

Starlight258 commented on src/main/java/attendance/util/TimeFormatter.java:
```diff
@@ -0,0 +1,34 @@
+    private static final DateTimeFormatter DATE_FORMATTER = DateTimeFormatter.ofPattern("MM월 dd일 E요일")
+            .withLocale(Locale.KOREAN);
+    private static final DateTimeFormatter TIME_FORMATTER = DateTimeFormatter.ofPattern("HH:mm");
+    private static final String DEFAULT_TIME_FORMAT = "--:--";
```
`TimeFormatter`가 출력 값에 대한 로직을 담고 있으므로, `InputValidator`, `MenuOption`과 마찬가지로 `view` 패키지로 옮기는 것이 좋을지 궁금해요!

EastHShin replied:
네 저도 단순 유틸성 클래스이기 보단, view에 대한 책임을 가지고 있는 클래스라 생각이 들어 View 패키지에 있는 것이 더 자연스럽다 생각되네요!

Starlight258 committed:
test(CrewInitializer): 크루 초기화 테스트 작성

Starlight258 committed:
test(SubjectType): 테스트 추가

Starlight258 committed:
refactor(CrewInitializer): 필요없는 코드 제거

- Map으로 데이터를 받아오기 때문에 중복된 이름(key)가 들어올 수 없다.

Starlight258 committed:
refactor(SubjectType, AttendanceType): enum에서 view의 로직 제거

- 한글 상수명은 view의 로직이므로 view에서 관리하도록 수정한다.

Starlight258 committed:
refactor(StringParser): 테스트 추가

Starlight258 committed:
refactor(CrewInitializer, ModifyCommand): 분산되어 있는 현재 시간 상수 하나로 모으기

- 분산되어 있는 현재 시간을 나타내는 2024, 12 매직 넘버를 LocalDate.now(clock)에서 가져오도록 수정한다.
 - 그에 따라 다른 해, 다른 달을 지원할 수 있으므로 확장성이 커진다.

Starlight258 committed:
refactor(AttendanceStatus, WarningLevel): enum 이름 명확하게 변경

Starlight258 committed:
refactor(ResultView): 상수에 final 붙이기

Starlight258 committed:
refactor(AttendanceWarning): dto 이름 변경

- dto 클래스명 끝에 dto를 붙이지 않아도 명확하기 때문에 수정한다.

Starlight258 committed:
refactor(All): 패키지 구조 변경

- domain 패키지 안에 model 패키지를 둘 필요가 없다. model 패키지를 삭제한다.

EastHShin commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,56 @@
+        );
+    }
+
+    private void process(final CrewHistories crewHistories, final MenuOption menuOption) {
+        Command command = commands.get(menuOption);
+        command.execute(crewHistories);
```
커맨트 패턴 사용으로 각 기능별로 책임이 분리가 되었네요! 👍

EastHShin commented on src/test/java/attendance/domain/AttendanceStatusTest.java:
```diff
@@ -0,0 +1,32 @@
+package attendance.domain;
```
AttendanceCounter 에 대한 테스트가 없네요!

Starlight258 replied:
리팩토링하면서 테스트 범위를 생각하지 못했네요!! 수정하도록 하겠습니다! 감사합니다.

EastHShin commented on src/test/java/attendance/fixture/TestFixture.java:
```diff
@@ -0,0 +1,29 @@
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+
+public class TestFixture {
```
TestFixture 사용 👍

EastHShin reviewed (approved):
안녕하세요 밍트!

깔끔하게 피드백 반영 잘 해주셨네요
이번 단계는 머지하고 step2 에서 뵙겠습니다. 고생하셨습니다 :)
